### PR TITLE
Edge-case handling: ships mit nur einer Pos.

### PR DESCRIPTION
**Pull request by phiros**
Mir ist aufgefallen, dass das Zeugs explodiert sobald man einen test mit Berta
inkludiert weil:
- travelTime ist dann immer 0.0 was bei allen `avgSpeedOf` Methoden in einem NaN Rückgabewert endet
- das pattern matching bei der Lösung mit sliding war nicht erschöpfend. Wenn nur eine Position
  im Vector ist gibt es eine (Runtime) Exception (Der Vector sieht dann so aus: `Vector(Vector(Position(...))))`.

Falls du Dich an nen paar Stellen wunderst: 
- Ich hab zwischendurch mal intelij gesagt es soll auto format machen...
- Die code duplizierung ist gewollt.
- Die andere Formatierung der if Anweisung bei der Imparative Lösung ist absichtlich so gewählt.

### CODE REVIEW

**Comment by Jentsch** [comment]
Vorweg, ich habe an diesen Fall nicht gedacht, als ich Code geschrieben habe. Mit ScalaSpecs wäre das nicht passiert, hatte mich aber dagegen entschieden, da so im Training die Tests besser zu lesen waren.

Ich finde NaN semantisch präziser als 0, weil die Geschwindigkeit wirklich nicht bekannt ist. Allerdings ist NaN immer wieder überraschend, weil viele an diesen Fall nicht denken. Um den Fall prominenter zu machen, könnte man auch Option[Double] verwenden, da weiß man was man hat – oder eben nicht. Ich behaupte, das alle drei Lösungen keinen Spaß machen und würde überlegen, ob man nicht die Vorbedingungen verschärft, also konkretere Typen verwendet.

Da hätte ich mir `sliding` wohl besser ansehen sollen. Würde es zusätzlich noch als Contra vermerken, dass der Compiler diesen Fehler nicht aufdecken konnte.

**Comment by phiros** [comment]
Hmmm Ja hatte auch daran gedacht das es eigentlich nicht ganz klar ist ob bei einem Weg-Zeit Datenpunkt die Geschwindigkeit überhaupt definiert ist. Ich fand gestern das 0 eigentlich ein ganz OKer  Wert ist um diesen Umstand zu repräsentieren (Die Idee war: alles hat Geschwindigkeit 0 solange nichts anderes bekannt ist). Bin mir jetzt nach deinem Einwand auch nicht mehr so sicher ob das so klug war (BTW: Option hatte ich auch überlegt habs dann aber gelassen).

Zu `sliding`: Da war ich auch überrascht. Allerdings hab ich jetzt auch verstanden wieso in dem Beispiel was die Leute vorgestellt haben `collect` verwendet wurde. Aus der Doku zu `Seq.collect`:

```
[use case]
Builds a new collection by applying a partial function to all elements of this sequence on which the function is defined.
```

**Comment by Jentsch** [comment]
Habe eine Branch “more-types” angelegt, in dem ich zum einem Type-Tagging verwendet habe, zum anderen Path-Dependent-Types. Kannst ja mal drauf schauen, ob es noch besser geht.

File: src/main/scala/com/example/PositionOfShips.scala [hunk 1]
```diff
@@ -28,12 +29,16 @@ class PositionOfShips(positions: Vector[Position]) {
     val totalDistance =
       pos
         .sliding(2)
-        .map { case Vector(a, b) =>
+        .map {
+        case Vector(a) => 0.0
```
Comment (Jentsch) [inline src/main/scala/com/example/PositionOfShips.scala:33]: Kannst Du diese Änderungen in einen extra Commit rebases? Dann würde ich denn schon pullen. (Und sehen, ob das überhaupt so geht, wie ich mir das vorstelle mit Git)

Comment (phiros) [inline src/main/scala/com/example/PositionOfShips.scala:33]: Welche änderungen meinst du?
Willst du alle Änderungen in `PositionOfShips.scala` in commits zerlegt haben
oder die Änderungen in `PositionOfShips.scala` in ein commit abgetrennt von
z.B. .gitignore änderungen?
